File: hack/integration/maintenance.py
"""Live same-pin restart and RetainData deletion in the harness-owned cluster.

Call only from run.py after its identity/network/storage fixtures pass. This
module never discovers a kubeconfig, starts infrastructure, or changes AWS.
"""
import json
import time


def exercise(k, get, wait_for, restart_operator, ready, curl):
    def reservation(fleet):
        values = json.loads(k('get', 'celldstoragereservations', '-o', 'json'))['items']
        return next(value for value in values if value['spec']['fleetName'] == fleet
                    and value['spec']['fleetNamespace'] == 'fleets')

    def journal(fleet):
        value = reservation(fleet)
        return json.loads(value['metadata']['annotations']['celld.example.com/lifecycle-journal'])

    def pods(uid):
        return json.loads(k('-n', 'fleets', 'get', 'pods', '-l',
                            'celld.example.com/fleet-uid=' + uid, '-o', 'json'))['items']

    for fleet, kind, probe in [('alpha', 'deployment', 'client'),
                               ('beta', 'statefulset', 'client-beta')]:
        uid = get('celldfleet', fleet)['metadata']['uid']
        k('-n', 'fleets', 'patch', 'celldfleet', fleet, '--type=merge', '-p',
          json.dumps({'spec': {'capacity': None, 'replicas': 3, 'maintenance': None}}))
        wait_for(lambda: get(kind, fleet)['spec']['replicas'] == 3 and ready(fleet)
                 and not journal(fleet).get('Operation'),
                 'maintenance fixture has three ready replicas: ' + fleet, timeout=360)
        assert json.loads(curl(probe, fleet, '/?cell=maintenance&id=ack', 8080,
                               method='PUT'))['stored'] is True
        before = {pod['metadata']['uid'] for pod in pods(uid)}
        claims = {claim['metadata']['name']: claim['metadata']['uid'] for claim in
                  json.loads(k('-n', 'fleets', 'get', 'pvc', '-l',
                               'celld.example.com/fleet-uid=' + uid, '-o', 'json'))['items']}
        token = 'live-maintenance-' + fleet
        k('-n', 'fleets', 'patch', 'celldfleet', fleet, '--type=merge', '-p',
          json.dumps({'spec': {'maintenance': {'restartToken': token}}}))
        wait_for(lambda: (journal(fleet).get('Maintenance') or {}).get('Phase') in
                 ('Stopping', 'Authorized', 'Recovering'),
                 'restart irreversible authority persisted: ' + fleet, timeout=180)
        restart_operator()
        wait_for(lambda: token in journal(fleet).get('CompletedRestarts', [])
                 and ready(fleet), 'restart recovers across operator replacement: ' + fleet,
                 timeout=600)
        after = {pod['metadata']['uid'] for pod in pods(uid)}
        assert len(after) == 3 and not before.intersection(after), (before, after)
        assert json.loads(curl(probe, fleet, '/?cell=maintenance&id=ack', 8080))['stored'] is True
        for name, claim_uid in claims.items():
            assert get('pvc', name)['metadata']['uid'] == claim_uid
        restart_operator()
        time.sleep(12)
        assert {pod['metadata']['uid'] for pod in pods(uid)} == after, 'completed token replayed'
        print('PASS: exact-UID rolling restart, acknowledged data, retained claims, and token replay: '
              + fleet, flush=True)

        retained = reservation(fleet)
        k('-n', 'fleets', 'delete', 'celldfleet', fleet, '--wait=false')
        def finalized():
            values = json.loads(k('-n', 'fleets', 'get', 'celldfleets', '-o', 'json'))['items']
            return all(value['metadata']['uid'] != uid for value in values)
        wait_for(finalized, 'RetainData finalizer completes with proven shutdown: ' + fleet,
                 timeout=600)
        assert not pods(uid)
        assert reservation(fleet)['metadata']['uid'] == retained['metadata']['uid']
        assert journal(fleet)['Maintenance']['Phase'] == 'Cleanup'
        for name, claim_uid in claims.items():
            assert get('pvc', name)['metadata']['uid'] == claim_uid
        print('PASS: final shutdown removes compute and preserves reservation/PVC identities: '
              + fleet, flush=True)

File: hack/integration/maintenance.go
package main

import (
	"fmt"
	"time"
)

// exerciseMaintenance qualifies rolling restart and runtime upgrade on
// retained disks (ADR 0023): one member at a time, highest ordinal first, no
// coordinated downtime, every PVC/PV/CSI identity unchanged and every
// acknowledged write readable. It ends with deletion of both profiles.
func (h *harness) exerciseMaintenance() {
	h.exercisePersistentRestart()
	h.exercisePersistentUpgrade()
	h.exerciseBucketRestart()
	h.deleteFleet("beta")
	h.deleteFleet("alpha")
}

// rollAll applies patch to a PersistentFleet and waits until every member
// has a new Pod on its old disk, watching the one-disruption invariant.
// midway, when set, runs once after the first member has been replaced.
func (h *harness) rollAll(fleetName, patch string, done func() bool, midway func()) {
	count := specReplicas(h.get("statefulset", fleetName))
	disks := h.claims(fleetName)
	before := nameUIDs(h.memberPods(fleetName))
	d := h.watchDisruptions(fleetName)
	h.merge(fleetName, patch)
	h.waitWatching("rolling "+fleetName+" on retained disks", 20*time.Minute, d, func() bool {
		if midway != nil && len(d.replaced) > 0 {
			midway()
			midway = nil
		}
		after := nameUIDs(h.memberPods(fleetName))
		for name, uid := range before {
			if after[name] == uid {
				return false
			}
		}
		return h.settled(fleetName, count) && (done == nil || done())
	})
	must(sameDisks(disks, h.claims(fleetName)))
	want := make([]string, 0, count)
	for i := count - 1; i >= 0; i-- {
		want = append(want, fmt.Sprintf("%s-%d", fleetName, i))
	}
	assert(same(d.replaced, want), "%s members were not replaced highest ordinal first: %v", fleetName, d.replaced)
	fmt.Println("PASS:", fleetName, "rolled one member at a time, highest ordinal first; every disk retained")
}

func (h *harness) exercisePersistentRestart() {
	h.scale("beta", 3)
	h.writeLedger("beta")
	h.startWriter("beta")
	before := nameUIDs(h.memberPods("beta"))
	token := "rolling-restart-beta"
	// A paused fleet accepts the token but changes nothing.
	h.merge("beta", encode(object{"spec": object{"maintenance": object{"paused": true, "restartToken": token}}}))
	h.hold(15*time.Second, "pause holds a requested restart", func() bool { return same(nameUIDs(h.memberPods("beta")), before) })
	h.rollAll("beta", `{"spec":{"maintenance":{"paused":false}}}`, nil, nil)
	h.stopWriter("beta")
	h.readLedger("beta")
	after := nameUIDs(h.memberPods("beta"))
	h.restartOperator()
	h.hold(20*time.Second, "completed restart token does not replay", func() bool { return same(nameUIDs(h.memberPods("beta")), after) })
}

func (h *harness) exercisePersistentUpgrade() {
	if h.opts.upgradeFrom == "" {
		fmt.Println("NOT RUN: runtime upgrade on retained disks; --upgrade-from none")
		return
	}
	legacy := h.newFleet("legacy", "bucket-legacy", "PersistentFleet", "fleets")
	legacy.Spec.RuntimeImage = h.opts.upgradeFrom
	legacy.Spec.Replicas = 3
	h.apply(legacy)
	h.waitFor("legacy runtime serves on retained disks", 8*time.Minute, func() bool {
		ready := 0
		for _, pod := range h.memberPods("legacy") {
			if podReady(pod) {
				ready++
			}
		}
		return h.ready("legacy") && ready == 3
	})
	reason, message := h.fleetReason("legacy")
	assert(reason != "Provisioned", "a runtime without node_log state reported a settled fleet: %s", message)
	fmt.Println("PASS: legacy runtime is served but never reported settled:", reason, message)
	h.writeLedger("legacy")
	h.startWriter("legacy")
	h.rollAll("legacy", encode(object{"spec": object{"runtimeImage": h.opts.runtimeImage}}), func() bool {
		for _, pod := range h.memberPods("legacy") {
			if str(list(pod, "spec", "containers")[0], "image") != h.opts.runtimeImage {
				return false
			}
		}
		return true
	}, nil)
	h.stopWriter("legacy")
	h.readLedger("legacy")
	fmt.Println("PASS: runtime upgrade", h.opts.upgradeFrom, "->", h.opts.runtimeImage, "kept every disk and acknowledged write")
	h.deleteFleet("legacy")
}

func (h *harness) exerciseBucketRestart() {
	h.scale("alpha", 2)
	h.writeLedger("alpha")
	h.startWriter("alpha")
	before := nameUIDs(h.memberPods("alpha"))
	h.merge("alpha", encode(object{"spec": object{"maintenance": object{"restartToken": "rolling-restart-alpha"}}}))
	h.waitFor("Ordered Bucket rolling restart", 10*time.Minute, func() bool {
		after := nameUIDs(h.memberPods("alpha"))
		for name, uid := range before {
			if after[name] == uid {
				return false
			}
		}
		return h.settled("alpha", 2)
	})
	assert(h.budget("alpha") == 1, "Bucket PDB does not allow one disruption")
	h.stopWriter("alpha")
	h.readLedger("alpha")
	fmt.Println("PASS: Ordered Bucket rolling restart replaced every member with acknowledged writes intact")
}
